### DISCUSSION TITLE
How to create a Scene.environmentTexture (before engine/scene is initialized) ?

### DISCUSSION THREAD

**Discussion by sevaseva** [brianzinn/react-babylonjs#271] [Q&A] [answered]
Hi, I hope everyone is doing well.

I'd like to set `<Scene ... environmentTexture={...}` attribute as I use a 'react-babylonjs' Scene component.

`new CubeTexture(...)` constructor that I use to create the texture accepts `sceneOrEngine` argument and actually seem to require a real one in order to work.

A naive attempt to pass either `useEngine()`  or `useScene()` fails to
```
Unhandled Runtime Error
TypeError: this._getEngine() is null
```
which makes sense as the CubeTexture constructor actually runs before the return statement is executed so there is no Engine yet.

How do I make something like this work?

-----------

More concrete short snippet of the relevant code in my component source file:
```jsx
import { Scene, Engine, useEngine, useCanvas, useScene } from 'react-babylonjs'

import {Vector3, CubeTexture, ...} from "@babylonjs/core"

export default function MyScene({ id }) {
    const env = new CubeTexture(
        "foo.env",
        useScene(),...);
    env.rotationY = ....; ...; ...;
    return <>
        <Engine
            antialias
            adaptToDeviceRatio
            canvasId={id}
            renderOptions={{
                whenVisibleOnly: true,
            }}
        >
            <Scene
                clearColor={Color3.Black()}
                ambientColor={Color3.Black()}
                environmentTexture={env}
            >
....
```

**Comment by brianzinn** [comment]
Does this help?
https://brianzinn.github.io/react-babylonjs/examples/hooks/scene-engine

You have also a callback when the scene is created. Also, a skybox component. Otherwise you have a chicken/egg problem.

**Reply by sevaseva** [reply]
Thanks for a prompt response, @brianzinn

> Does this help? https://brianzinn.github.io/react-babylonjs/examples/hooks/scene-engine

No. (obviously, right? because that example is about a child component and I am trying to set a value to an _attribute_ of a Scene)

More surprisingly the following also does **not** work:
```
function createEnvTexture(engine) {
    const result = new CubeTexture("foo.env", engine);
    result.rotationY = Math.PI;
    return result;
}

export default function MyScene() {
    return <>
        <Engine>
            <Scene environmentTexture={createEnvTexture(useEngine())}>
                <hemisphericLight
                    name="light1"
                    intensity={0.7}
                    direction={new Vector3(0, 1, 0)}
                />
            </Scene>
        </Engine>
    </>
}
```
At runtime engine is null inside the `new CubeTexture` call. Note that useEngine() is called _from within_ the Engine "tag". For that reason I expect useEngine()'s return value to be an initialized engine (because that would be consistent with useScene() behavior documented in https://brianzinn.github.io/react-babylonjs/examples/hooks/scene-engine ) but it is a null. This seems surprising, seems to be at least a missing feature (if not a API design bug - depending how you look at it). Am I missing something? I'm happy to file an issue for this if you want me to.


> You have also a callback when the scene is created.

Are you referring to `onSceneMount` or some other callback?


> Also, a skybox component.

The following use of Skybox component worked to attach and render the right texture in the scene in some way (I will double-check if it's rendered the same exact way as assigning `environmentTexture` attribute at runtime, and go from there):
```
<Scene>
<Skybox rootUrl='foo.env' rotationY={Math.PI} boundingBoxSize={new Vector3(3.6, 1.6, 3.5} boundingBoxPosition={Vector3.Zero()} />
...
</Scene>
```

---------------------

> I don’t know if I implemented this:
> https://forum.babylonjs.com/t/environment-texutre-doesnt-work-when-develop-with-react-babylonjs/20936

That seems to be a negative. I verified by adding that `assignTo="environmentTexture"` argument to the texture and observing that the texture is loaded (appears as a texture in the inspector on the resulting page) but environmentTexture isn't set (according to the inspector on the page) and the texture isn't rendered (as opposed to when I use a Skybox child component).

I don't have a strong opinion as to whether `<cubeTexture ...  assignTo="environmentTexture" />` syntax/API you proposed there is good (it seems reasonable but I don't know the codebase well enough to anticipate possible downsides. I can file an issue for implementing that if you think it's good enough.

Speaking of syntax/API. Given that
1. I don't see any reasonable way for `<Scene ... environmentTexture={...}>` attribute to actually work due to the chicken and egg problem (unless `useEngine()` behavior is fixed and then it can be used to create the texture there as in the example I showed above?...) and
2. a user can very easily discover that attribute (with code complete, etc), naturally decide to use it, and then lose lots of time trying to make it work (as I did) before they find there is actually _no_ way to make it work,

the question arises:
* should that attribute be dropped entirely from Scene component (because it is misleading and can easily lead to users losing time)
* OR (if perhaps it has to be there because that's the data model and it's used at runtime, etc) can it be made more obvious (at compile time or runtime, _not_ only in the documentation ... because who reads documentation, right?) that this attribute is not for setting from between `<> </>` because of that chicken and egg problem?

**Reply by brianzinn** [reply]
it does work if you assign it a texture.  ie:
```jsx
const MyScene = () => {

  const engine = useEngine();
  const [textureReady, setTextureReady] = useState(false);
  const cubeTextureRef = useRef<undefined | CubeTexture>();

  useEffect(() => {
    if (engine !== null) {
      console.log('engine', engine);
      cubeTextureRef.current = new CubeTexture('../../../assets/textures/environment.env', engine);
      setTextureReady(true); // force a re-render to attach
      console.log('texture:', cubeTextureRef.current);
    }
  }, [engine])

  return (
    <Scene environmentTexture={cubeTextureRef.current}>
      ...
    </Scene>
  )
}
```

**Reply by sevaseva** [reply]
Thank you for this example, @brianzinn , it is a good one for me to learn.

However, while you are welcome to disagree, but I am pretty convinced that ... if what you just demonstrated is really the _intended_ way of using Scene.environmentTexture attribute in react-babylonjs, then it's a pretty darn terrible corner of react-babylonjs API. It is hard for me to imagine an engineer who, seeing this environmentTexture attribute supported by Scene component and knowing what environmentTexture is in babylonjs, writes something like you just wrote in a first/second/third attempt to set the env texture in their scene.

I don't mean to attack you. And I know you will be first to admit that this library (as well as every other library) isn't perfect, and it isn't for every use case or every user.

But yeah, while it is nice to know environmentTexture attribute isn't completely useless, and it can be put to use if one knows what they are doing and they for some reason _really_ prefer setting this attribute to using Skybox component,.. after seeing this example I am still of the opinion that it'd be better (for the quality of the API and overall usability of react-babylonjs) if this attribute didn't exist at all (or wasn't visible to users).

Because I do like some nice things this library provides (like `<Engine renderOptions={{ whenVisibleOnly: true }}>` that works and I can trust that you implemented that better and cleaner than I probably would), I'll leave my usage to minimum that brings me specific benefits I want, but do most of the things from a looooong callback function where I'll just code against babylonjs APIs.

If at some point I'll see a tangible benefit for my codebase from building scenes using declarative react-babylonjs (rather than babylonjs APIs directly) I'll consider it again.

Thank you very much. Keep up good work!

**Reply by brianzinn** [reply]
i put up the sample to help the next person:
https://brianzinn.github.io/react-babylonjs/examples/basic/assign-to

no offense, but your example code would create a new texture with every render
```jsx
// not a good practice:
<Scene environmentTexture={createEnvTexture(useEngine())}
```

if you're not going declarative then you are better off with `babylonjs-hook`  https://www.npmjs.com/package/babylonjs-hook, although not sure if all the same render options made their way over to there....

**Reply by sevaseva** [reply]
Cool, thank you very much!

**Reply by sevaseva** [reply]
(cannot mark this as the new Answer though because it's not a top level response)

**Comment by brianzinn** [comment]
I don’t know if I implemented this:

https://forum.babylonjs.com/t/environment-texutre-doesnt-work-when-develop-with-react-babylonjs/20936

**Comment by brianzinn** [comment, marked as answer]
I got this working:
```jsx
<Engine antialias adaptToDeviceRatio canvasId="babylon-js">
        <Scene>
          <arcRotateCamera
            name="camera1"
            target={Vector3.Zero()}
            minZ={0.001}
            alpha={0}
            beta={Math.PI / 2}
            radius={5}
            lowerBetaLimit={2}
            upperRadiusLimit={5}
          />
          <hemisphericLight name="light1" intensity={0.9} direction={Vector3.Down()} />
          <sphere name='sphere1' segments={16} diameter={2}>
            <pbrMaterial
              name='pbr'
              albedoColor={new Color3(1, 0.766, 0.336)}
              metallic={1.0}
              roughness={1.0}
              useRoughnessFromMetallicTextureAlpha={false} useRoughnessFromMetallicTextureGreen useMetallnessFromMetallicTextureBlue>
              <texture url='../../assets/textures/mr.jpg' assignTo='metallicTexture' />
            </pbrMaterial>
          </sphere>
          <cubeTexture level={0.5} assignTo="environmentTexture" rootUrl='../../assets/textures/environment.env' />
        </Scene>
      </Engine>
```
![image](https://user-images.githubusercontent.com/542756/216203313-b2188c23-cd64-4d6e-b796-a2960dbb3076.png)

Would that work for you? I'll need to commit the code changes before it would work on your side.

**Reply by brianzinn** [reply]
this was the example that I copied: https://playground.babylonjs.com/#2FDQT5#1153

**Reply by sevaseva** [reply]
> Would that work for you?

I assume you mean that the change you drafted adds `assignTo` attribute to _all_ texture components including `cubeTexture`, `texture` and others, and the expected behavior is that it
1. fails at compile time or runtime if the value assigned to that attribute is not a string (because we interpret that value as an attribute in a component)
2. assigns the texture[1] as value to the attribute of the _immediate parent_ component
3. fails at compile time or runtime if the attribute with that name - like `metallicTexture` or `environmentTexture` - isn't found in the immediate parent component OR if assigning the texture[1] to that attribute produces an compilation or runtime error

[1] I don't know how textures (values like `import {CubeTexture} from "@babylonjs/core"; const x = new CubeTexture(...);`) correspond to react components like `cubeTexture` or `texture` of `react-babylonjs`, but I hope that you clarify that in comments in the code and documentation you add with this change.

If that's what you have implemented, it looks like a reasonable change to me.

-----

Also adding `import { Skybox } from 'react-babylonjs'` component as a direct child of Scene worked for me. And I am actually curious exactly what's the difference - if any - there will be in the runtime behavior between using Skybox and using `<cubeTexture assignTo="environmentTexture">` direct child component of Scene. If you would mention that in comments and/or documentation that you commit with the code change and point me to that commit, I'd appreciate that.

(Actually, if there happens to be _no_ difference at all, then - depending on your philosophy for this library of components, it may not necessarily be a good idea to add this `assignTo` attribute new behavior. I haven't been looking at this library enough to know what this philosophy is.)

-----

Also another idea, now that you are adding this `assignTo` attribute (I am assuming no components currently have that attrubute and you are planning to add it only to texture components), consider just adding it to _all_ react-babylonjs components. Assuming that you can specify a relatively simple rule how the value to be assigned (e.g. `new CubeTexture(...)` value in my example) is deducted from any component. If this is easily specifiable without too much confusion and if it is implementable, it may be a better idea than just implement it as a special case only for textures (which I assume you are planning to do; although I suppose this may be a wrong assumption).

-------

Thanks! Please link the commit when you push it and let me know what version it will be released with. Keep up the good work!

**Reply by sevaseva** [reply]
I marked this response of yours as Answer. I am assuming you will commit that change. I'll mark something else as answer if you change your mind. Keep us posted :)

**Comment by sevaseva** [comment]
... on the other hand if you add `<cubeTexture ...  assignTo="environmentTexture" />` it means you plan to keep Scene.environmentTexture attribute.

Are you planning to address in some way the issue of Scene.environmentTexture being present and easily discoverable, but attempts to use it naturally leading to frustration and lost time?
